tensai-api/src/main/java/dev/vmsa/tensai/utils/EventEmitter.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * This file is part of tensai, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) 2022 VMSA
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package dev.vmsa.tensai.utils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * <p>Emit and listen for events. You can transform event outputs with {@link #then(Function)} method, or simply
+ * listen for events with {@link #listen(Consumer)}.</p>
+ * <p><b>Listeners chaining: </b> You can chain multiple listeners with {@link #then(Function)}: <pre>
+ * {@code
+ * // Initial event
+ * public static final EventEmitter<String> MY_EVENT = new EventEmitter<>();
+ * // 1st pass
+ * public static final EventEmitter<Integer> MY_EVENT_INT = MY_EVENT.then(s -> Integer.parse(s));
+ * public static final EventEmitter<Double> MY_EVENT_DOUBLE = MY_EVENT.then(s -> Double.parse(s));
+ * // 2nd pass
+ * public static final EventEmitter<Boolean> IS_ANSWER = MY_EVENT_INT.then(i -> i == 42);
+ * }</pre></p>
+ *
+ * @param <T> Type of event object.
+ */
+public class EventEmitter<T> {
```

Comment:
Can be replaced by 3rd lib (compatibility)

Reply:
Not sure which 3rd party library are we going to use. I'm trying to make Tensai as portable as possible, such as not relying on 3rd libraries or extra plugins/mods,

Which library you want to use?

Reply:
KyoriPowered/bus

Reply:
Im fully awarded the situation; however, I reckon that we should use platform's implementation instead.